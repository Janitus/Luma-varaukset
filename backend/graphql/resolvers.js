const { UserInputError, AuthenticationError } = require('apollo-server-errors')
const { readMessage } = require('../services/fileReader')
const { findValidTimeSlot, calculateAvailabelTimes, calculateNewTimeSlot, formatAvailableTimes } = require('../utils/timeCalculation')
const bcrypt = require('bcrypt')
const jwt = require('jsonwebtoken')
const mailer = require('../services/mailer')
const config = require('../utils/config')

const User = require('../models/user')
const Event = require('../models/event')
const Visit = require('../models/visit')
const Extra = require('../models/extra')
const Tag = require('../models/tag')
const Form = require('../models/forms')
const FormSubmissions = require('../models/formSubmissions')
const { addNewTags } = require('../utils/helpers')

const resolvers = {
  Query: {
    getUsers: async () => {
      const users = await User.find({})
      return users
    },
    getEvents: async () => {
      const date = new Date()
      date.setDate(date.getDate() - 90)
      const events = await Event.find({ end: { $gt: date.toISOString() } })
        .populate('tags', { name: 1, id: 1 })
        .populate('visits')
        .populate('extras')
      return events
    },
    getTags: async () => {
      const tags = await Tag.find({})
      return tags
    },
    getVisits: async (root, args, { currentUser }) => {
      if (!currentUser) {
        return []
      }
      const visits = await Visit.find({})
        .populate('event', { id: 1, title: 1, resourceids: 1, remoteVisit: 1, inPersonVisit : 1 })
        .populate('extras')
      return visits
    },
    findVisit: async (root, args) => {
      try {
        const visit = await Visit.findById(args.id).populate('extras')
        return visit
      } catch (e) {
        throw new UserInputError('Varausta ei lÃ¶ytynyt!')
      }
    },
    me: (root, args, context) => {
      return context.currentUser
    },
    getExtras: async () => {
      const extras = await Extra.find({})
      return extras
    },
    getForm: async (root, args) => {
      try {
        const form = await Form.findById(args.id)
        return form
      } catch (e) {
        throw new UserInputError('Form not found!')
      }
    },
    getForms: async () => {
      const forms = await Form.find({})
      return forms
    },
    getFormSubmissions: async (root, args, { currentUser }) => {
      if (!currentUser) {
        throw new AuthenticationError('not authenticated or no credentials')
      }
      try {
        const formSubmissions = await Form.findById(args.id).populate('submissions')
        return formSubmissions
      } catch (error) {
        throw new UserInputError('Form submissions not found!')
      }
    },
    getFormSubmission: async (root, args, { currentUser }) => {
      if (!currentUser) {
        throw new AuthenticationError('not authenticated or no credentials')
      }
      try {
        const formValues = await FormSubmissions.findById(args.id)
        return formValues
      } catch (e) {
        throw new UserInputError('Form submission not found!')
      }
    }
  },
  Visit: {
    event: async (root) => {
      const event = await Event.findById(root.event).populate('tags', { name: 1, id: 1 }).populate('extras')
      return event
    },
  },
  Form: {
    fields: (form) => JSON.stringify(form.fields)
  },
  FormSubmissions: {
    values: (submission) => JSON.stringify(submission.values)
  },
  Mutation: {
    createUser: async (root, args, { currentUser }) => {
      if (!currentUser || currentUser.isAdmin !== true) {
        throw new AuthenticationError('not authenticated or no credentials')
      }
      if (args.username.length < 5) {
        throw new UserInputError('username too short')
      }
      const salt = 10
      const passwordHash = await bcrypt.hash(args.password, salt)
      const newUser = new User({
        username: args.username,
        passwordHash,
        isAdmin: args.isAdmin,
      })
      await newUser.save()
      return newUser
    },
    login: async (root, args) => {
      const user = await User.findOne({ username: args.username })
      const passwordCorrect = user === null
        ? false
        : await bcrypt.compare(args.password, user.passwordHash)
      if (!(user && passwordCorrect)) {
        throw new UserInputError('Wrong credentials!')
      }
      const userForToken = { username: user.username, id: user._id }
      return { value: jwt.sign(userForToken, config.SECRET) }
    },
    createEvent: async (root, args, { currentUser }) => {
      if (!currentUser) throw new AuthenticationError('not authenticated')
      if (args.grades.length < 1) throw new UserInputError('At least one grade must be selected!')
      if (args.title.length < 5)  throw new UserInputError('title too short')

      const mongoTags = await addNewTags(args.tags)

      const extras = await Extra.find({ _id: { $in: args.extras } })

      const newEvent = new Event({
        title: args.title,
        start: args.start,
        end: args.end,
        desc: args.desc,
        resourceids: args.scienceClass,
        grades: args.grades,
        remotePlatforms: args.remotePlatforms,
        otherRemotePlatformOption: args.otherRemotePlatformOption,
        remoteVisit: args.remoteVisit,
        inPersonVisit: args.inPersonVisit,
        waitingTime: args.waitingTime,
        availableTimes: [{
          startTime: args.start,
          endTime: args.end
        }],
        duration: args.duration,
        customForm: args.customForm
      })

      newEvent.extras = extras
      newEvent.tags = mongoTags
      await newEvent.save()
      return newEvent
    },
    modifyEvent: async (root, args, { currentUser }) => {
      const extras = await Extra.find({ _id: { $in: args.extras } })
      const { title, desc, resourceids, grades, remotePlatforms, otherRemotePlatformOption, remoteVisit, inPersonVisit, customForm } = args
      const event = await Event.findById(args.event).populate('visits')
      if (!currentUser) throw new AuthenticationError('not authenticated')
      if (new Date(args.start).getHours() < 8) throw new UserInputError('invalid start time')
      if (new Date(args.end).getHours() > 17) throw new UserInputError('invalid end time')

      title !== undefined ? event.title = title : null
      desc !== undefined ? event.desc = desc : null
      resourceids !== undefined ? event.resourceids = resourceids : null
      grades !== undefined ? event.grades = grades : null
      remotePlatforms !== undefined ? event.remotePlatforms = remotePlatforms : null
      otherRemotePlatformOption !== undefined ? event.otherRemotePlatformOption = otherRemotePlatformOption : null
      remoteVisit !== undefined ? event.remoteVisit = remoteVisit : null
      inPersonVisit !== undefined ? event.inPersonVisit = inPersonVisit : null
      customForm !== undefined ? event.customForm = customForm : null
      event.extras = extras
      event.tags = await addNewTags(args.tags)
      if (event.visits.length) {
        const newTimeSlot = calculateNewTimeSlot(
          event.visits,
          args.start ? new Date(args.start) : new Date(event.start),
          args.end ? new Date(args.end) : new Date(event.end)
        )
        if (newTimeSlot) {
          const eventTimes = {
            start: new Date(newTimeSlot.start),
            end: new Date(newTimeSlot.end)
          }
          const newAvailableTimes = calculateAvailabelTimes(event.visits, eventTimes, event.waitingTime, event.duration)
          event.start = newTimeSlot.start.toISOString()
          event.end = newTimeSlot.end.toISOString()
          event.availableTimes = formatAvailableTimes(newAvailableTimes)
        } else {
          throw new UserInputError('invalid start or end')
        }
      } else {
        args.start ? event.start = args.start : null
        args.end ? event.end = args.end : null
        event.availableTimes = [{
          startTime: args.start ? args.start : event.start,
          endTime: args.end ? args.end : event.end
        }]
      }
      await event.save()
      return event
    },
    createVisit: async (root, args, { currentUser }) => {
      const event = await Event.findById(args.event).populate('visits', { startTime: 1, endTime: 1 })
      const visitTime = {
        startTime: new Date(args.startTime),
        endTime: new Date(args.endTime)
      }
      const availableTimes = event.availableTimes.map(time => ({
        startTime: new Date(time.startTime),
        endTime: new Date(time.endTime)
      }))
      const eventTime = {
        start: new Date(event.start),
        end: new Date(event.end)
      }
      const visitTimes = event.visits.concat(visitTime)

      if (!findValidTimeSlot(availableTimes, visitTime)) {
        throw new UserInputError('Given timeslot is invalid')
      }

      event.availableTimes = calculateAvailabelTimes(visitTimes, eventTime, event.waitingTime, event.duration)

      const visit = new Visit({
        ...args,
        event: event,
        status: true,
        extras: []
      })
      const extras = await Extra.find({ _id: { $in: args.extras } })
      visit.extras = extras

      let savedVisit
      event.availableTimes = formatAvailableTimes(event.availableTimes)
      try {
        const now = new Date()
        const start = new Date(event.start)
        const startsAfter14Days = start - now >= 1209600000
        const startsAfter1Hour = start - now >= 3600000
        const eventCanBeBooked = !currentUser ? startsAfter14Days : startsAfter1Hour
        if (eventCanBeBooked) {
          savedVisit = await visit.save()
          const details = [{
            name: 'link',
            value: `${config.HOST_URI}/${savedVisit.id}`
          }]
          const text = await readMessage('welcome.txt', details)
          const html = await readMessage('welcome.html', details)
          mailer.sendMail({
            from: 'Luma-Varaukset <noreply@helsinki.fi>',
            to: visit.clientEmail,
            subject: 'Tervetuloa!',
            text,
            html
          })
          event.visits = event.visits.concat(savedVisit._id)
          await event.save()
          return savedVisit
        }
      } catch (error) {
        event.availableTimes = availableTimes
        await event.save()
        await savedVisit.delete()
        throw new UserInputError(error.message, {
          invalidArgs: args,
        })
      }
    },
    cancelVisit: async (root, args) => {
      const visit = await Visit.findById(args.id)
      if (!visit || visit.status === false) throw new UserInputError('Varausta ei lÃ¶ydy')
      const event = await Event.findById(visit.event).populate('visits', { startTime: 1, endTime: 1 })
      const oldAvailableTimes = event.availableTimes.map(time => ({
        startTime: new Date(time.startTime),
        endTime: new Date(time.endTime)
      }))
      const visitTimes = event.visits.filter(v => v.id !== visit.id)

      const eventTime = {
        start: new Date(event.start),
        end: new Date(event.end)
      }
      let newAvailTimes = calculateAvailabelTimes(visitTimes, eventTime, event.waitingTime, event.duration)

      newAvailTimes = newAvailTimes.map(time => {
        if (typeof time.startTime === 'string') return time
        return {
          startTime: time.startTime.toISOString(),
          endTime: time.endTime.toISOString()
        }
      })

      try {
        event.visits = event.visits.filter(v => v.id.toString() !== visit.id)
        event.availableTimes = newAvailTimes
        visit.status = false
        await visit.save()
        await event.save()
        return visit
      } catch (error) {
        event.visits = event.visits.concat(visit._id)
        event.availableTimes = oldAvailableTimes
        visit.status = true
        await event.save()
        await visit.save()
        throw new UserInputError(error.message, {
          invalidArgs: args,
        })
      }
    },
    createExtra: async (root, args, { currentUser }) => {
      if (!currentUser) {
        throw new AuthenticationError('not authenticated or no credentials')
      }
      if(args.name.length < 3){
        throw new UserInputError('Name too short!')
      }
      if(args.classes.length === 0) {
        throw new UserInputError('Select at least one science class!')
      }
      if(args.remoteLength === 0 && args.inPersonLength === 0){
        throw new UserInputError('Give duration for at least one mode!')
      }
      try {
        const newExtra = new Extra({
          ...args
        })
        const savedExtra = await newExtra.save()
        return savedExtra
      } catch (error) {
        throw new UserInputError(error.message, { invalidArgs: args })
      }
    },
    deleteExtra: async (root, args, { currentUser }) => {
      if (!currentUser) {
        throw new AuthenticationError('not authenticated or no credentials')
      }
      if (!args.id) {
        throw new UserInputError('No ID provided!')
      }
      try {
        await Extra.deleteOne({ _id:args.id })
        return 'Deleted Extra with ID ' + args.id
      } catch (error) {
        throw new UserInputError('Backend problem')
      }
    },
    deleteEvent: async (root, args, { currentUser }) => {
      if (!currentUser) {
        throw new AuthenticationError('not authenticated or no credentials')
      }
      if (!args.id) {
        throw new UserInputError('No ID provided!')
      }
      try {
        const event = await Event.findById(args.id)
        if (event.visits.length) {
          throw new UserInputError('Event has visits!')
        }
        await Event.deleteOne({ _id:args.id })
        return 'Deleted Event with ID ' + args.id
      } catch (error) {
        throw new UserInputError('Event has visits!')
      }
    },
    deleteUser: async (root, args, { currentUser }) => {
      if (!currentUser || !currentUser.isAdmin) {
        throw new AuthenticationError('not authenticated or no admin priviledges')
      }
      if (args.id === currentUser.id) {
        throw new UserInputError('This user cannot be removed')
      }
      if (!args.id) {
        throw new UserInputError('No ID provided!')
      }
      try {
        await User.deleteOne({ _id: args.id })
        return 'Deleted user with ID ' + args.id
      } catch (error) {
        throw new UserInputError('User deletion failed')
      }
    },
    createForm: async (root, args, { currentUser }) => {
      if (!currentUser) {
        throw new AuthenticationError('not authenticated')
      }
      try {
        const newForm = new Form({
          name: args.name,
          fields: JSON.parse(args.fields)
        })
        const savedForm = await newForm.save()
        return savedForm
      } catch (error) {
        throw new UserInputError(error.message, { invalidArgs: args })
      }
    },
    updateForm: async (root, args, { currentUser }) => {
      if (!currentUser) {
        throw new AuthenticationError('not authenticated')
      }
      try {
        const form = await Form.findById(args.id)
        form.name = args.name
        form.fields = JSON.parse(args.fields)
        form.markModified('fields')
        await form.save()
        return form
      } catch (error) {
        throw new UserInputError(error.message, { invalidArgs: args })
      }
    },
    deleteForm: async (root, args, { currentUser }) => {
      if (!currentUser) {
        throw new AuthenticationError('not authenticated')
      }
      try {
        await Form.deleteOne({ _id: args.id })
        return 'Deleted form with ID ' + args.id
      } catch (error) {
        throw new UserInputError('Form deletion failed')
      }
    },
    createFormSubmission: async (root, args) => {
      try {
        const form = await Form.findById(args.formID)
        const newFormSubmissions = new FormSubmissions({
          form,
          values: JSON.parse(args.values)
        })
        const savedFormSubmissions = await newFormSubmissions.save()
        return savedFormSubmissions
      } catch (error) {
        throw new UserInputError(error.message, { invalidArgs: args })
      }
    }
  }
}

module.exports = resolvers