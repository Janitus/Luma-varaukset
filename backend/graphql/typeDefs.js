const { gql } = require('apollo-server-express')

//Visit- ja Event-tyyppien kent√§t gradeId ja online tarkistettava
const typeDefs = gql`
  type Lock {
    event: ID!
    token: String!
    locked: Boolean!
  }
  type Tag {
    id: ID
    name: String!
  }
  type User {
    id: ID!
    username: String!
    isAdmin: Boolean!
  }
  type TimeSlot {
    startTime: String!,
    endTime: String!
  }
  type Extra {
    id: ID!
    name: String!
    classes: [Int]!
    remoteLength: Int!
    inPersonLength: Int!
  }
  type Event {
    id: ID!
    title: String!
    resourceids: [Int]
    remoteVisit: Boolean!
    inPersonVisit: Boolean!
    grades: [Int]!
    remotePlatforms: [Int]
    otherRemotePlatformOption: String
    start: String!
    end: String!
    desc: String
    tags: [Tag]!
    visits: [Visit]!
    availableTimes: [TimeSlot]!
    waitingTime: Int!
    extras: [Extra]
    duration: Int!
    customForm: ID
    disabled: Boolean!
    locked: Boolean!
  }
  type Visit {
    id: ID!
    event: Event
    clientName: String!
    schoolName: String!
    schoolLocation: String!
    clientEmail: String!
    clientPhone: String!
    grade: String!
    participants: Int!
    extras: [Extra]
    status: Boolean!
    startTime: String!
    endTime: String!
    inPersonVisit: Boolean!
    remoteVisit: Boolean!
    dataUseAgreement: Boolean!
    remotePlatform: String
  }
  type Token {
    value: String!
  }
  type Form {
    id: ID!
    name: String!
    fields: String!
  }
  type FormSubmissions {
    id: ID
    form: Form
    values: String
  }
  type Query {
    getUsers: [User]!
    me: User!
    getEvents: [Event]!
    findVisit(id: ID!): Visit
    getVisits: [Visit]
    getTags: [Tag]!
    getExtras: [Extra]!
    getForm(id: ID): Form
    getForms: [Form]
    getFormSubmissions(formID: ID): FormSubmissions
    getFormSubmission(id: ID): FormSubmissions
  }
  type Mutation {
    lockEvent(
      event: ID!
    ): Lock
    unlockEvent(
      event: ID!
    ): Event
    resetPassword(
      user: ID!
      password: String!
    ): User
    changeUsername(
      user: ID!
      username: String!
      isAdmin: Boolean!
    ): User
    createUser(
      username: String!
      password: String!
      isAdmin: Boolean!
    ): User
    login(
      username: String!
      password: String!
    ): Token
    createEvent(
      title: String!
      scienceClass: [Int]!
      grades: [Int]!
      remotePlatforms: [Int]
      otherRemotePlatformOption: String
      start: String!
      end: String!
      desc: String
      inPersonVisit: Boolean!
      remoteVisit: Boolean!
      tags: [String]
      waitingTime: Int!
      extras: [ID]
      duration: Int!
      customForm: ID
    ): Event
    modifyEvent(
      event: ID!
      title: String
      resourceids: [Int]
      grades: [Int]
      remotePlatforms: [Int]
      otherRemotePlatformOption: String
      desc: String
      inPersonVisit: Boolean
      remoteVisit: Boolean
      extras: [ID]
      start: String
      end: String
      tags: [String]
      #waitingTime: Int!
      #duration: Int!
      customForm: ID
    ): Event
    createVisit(
      event: ID!
      clientName: String!
      schoolName: String!
      schoolLocation: String!
      clientEmail: String!
      clientPhone: String!
      startTime: String!
      endTime: String!
      grade: String!
      participants: Int!
      extras: [ID]
      inPersonVisit: Boolean!
      remoteVisit: Boolean!
      dataUseAgreement: Boolean!
      remotePlatform: String
      token: String!
    ): Visit
    disableEvent(
      event: ID!
    ): Event
    enableEvent(
      event: ID!
    ): Event
    cancelVisit(id: ID!): Visit
    createExtra(
      name: String!
      classes: [Int]!
      remoteLength: Int!
      inPersonLength: Int!
    ): Extra
    deleteExtra(
      id: String!
    ): String
    deleteEvent(
      id: String!
    ): String
    deleteUser(
      id: String!
    ): String
    createForm(
      name: String!
      fields: String
    ): Form
    createFormSubmission(
      formID: ID
      values: String
    ): FormSubmissions
    updateForm(
      id: ID!
      name: String!
      fields: String
    ): Form
    deleteForm(
      id: ID!
    ): String
  }
  type Subscription {
    eventModified: Event
  }
`
module.exports = typeDefs