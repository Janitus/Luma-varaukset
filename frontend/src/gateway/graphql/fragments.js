import { gql } from '@apollo/client'

export const EXTRA_FIELDS = gql`
  fragment ExtraFields on Extra {
    id
    name
    classes
    remoteLength
    inPersonLength
  }
`

export const EVENT_FIELDS = gql`
  ${EXTRA_FIELDS}
  fragment EventFields on Event {
    id
    title
    resourceids
    grades {
      name
    }
    remotePlatforms
    otherRemotePlatformOption
    waitingTime
    tags {
      id
      name
    }
    start
    end
    desc
    inPersonVisit
    remoteVisit
    schoolVisit
    availableTimes {
      startTime
      endTime
    }
    visits {
      startTime
      endTime
    }
    extras { ...ExtraFields }
    duration
    customForm {
      id
      name
      fields
    }
    disabled
    locked
    group {
      id
      name
      disabled
    }
    publishDate
    languages
    limits
    closedDays
  }
`

export const VISIT_FIELDS = gql`
  ${EXTRA_FIELDS}
  fragment VisitFields on Visit {
    id
    clientName
    schoolName
    schoolLocation
    clientEmail
    clientPhone
    event {
      id
      title
      start
      end
      resourceids
    }
    extras { ...ExtraFields }
    grade
    gradeInfo
    startTime
    endTime
    participants
    teaching {
      type
      location
      payload
    }
    status
    customFormData
    language
    created
    cancellation
  }
`
