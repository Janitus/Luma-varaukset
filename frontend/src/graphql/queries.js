import { gql } from '@apollo/client'

export const LOGIN = gql`
  mutation login($username: String!, $password: String!) {
    login(
      username: $username,
      password: $password
    ) {
      value
    }
  }
`

export const CREATE_USER = gql`
  mutation createUser($username: String!, $password: String!, $isAdmin: Boolean!) {
    createUser(
      username: $username,
      password: $password,
      isAdmin: $isAdmin
    ) {
      username,
      isAdmin
    }
  }
`

export const USERS = gql`
  query getUsers {
    getUsers {
      username,
      isAdmin
    }
  }
`

export const EVENTS = gql`
  query getEvents {
    getEvents {
      id
      title
      resourceId
      grades
      tags {
        id
        name
      }
      start
      end
      desc
      booked
    }
  }
`

export const TAGS = gql`
  query getTags {
    getTags {
      name
      id
    }
  }
`

export const CURRENT_USER = gql`
  query me {
    me {
      username,
      isAdmin
    }
  }
`

export const CREATE_EVENT = gql`
  mutation createEvent($title: String!, $start: String!, $end: String!, $scienceClass: String!, $grades: [Int]!, $remoteVisit: Boolean!, $closeVisit: Boolean!, $desc: String, $tags: [TagInput]) {
    createEvent (
      title: $title,
      start: $start,
      end: $end,
      class: $scienceClass,
      desc: $desc,
      grades: $grades,
      remoteVisit: $remoteVisit,
      closeVisit: $closeVisit,
      tags: $tags
    ) {
      id
      title
      resourceId
      grades
      start
      end
      tags {
        name,
        id
      }
      closeVisit
      remoteVisit
      desc
      booked
    }
  }
`

export const CREATE_VISIT = gql`
  mutation createVisit($event: ID!, $grade: Int!, $clientName: String!, $clientEmail: String!, $clientPhone: String!) {
    createVisit(
      event: $event,
      grade: $grade,
      clientName: $clientName,
      clientEmail: $clientEmail,
      clientPhone: $clientPhone
    ) {
      id
      event {
        id
        title
        booked
      }
      grade
      clientName
      clientEmail
      clientPhone
    }
  }
`
export const FIND_VISIT = gql`
  query findVisit($id: ID!) {
    findVisit(id: $id) {
      clientName
      clientEmail
      clientPhone
      event {
        title
        resourceId
        start
        end
      }
      grade
    }
  }
`

export const CANCEL_VISIT = gql`
  mutation cancelVisit($id: ID!) {
    cancelVisit(id: $id) {
      id
    }
  }
`