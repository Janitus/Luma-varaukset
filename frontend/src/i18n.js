import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'
import fi from './locales/fi/translations.json'
import en from './locales/en/translations.json'
import sv from './locales/sv/translations.json'

export const resources = {
  fi: {
    translation: fi,
    language: 'fi'
  },
  en: {
    translation: en,
    language: 'en'
  },
  sv: {
    translation: sv,
    language: 'sv'
  }
}

i18n
  .use(initReactI18next)
  .init({
    resources,
    fallbackLng: 'fi',
    debug: false,
    interpolation: {
      escapeValue: false,
    }
  })

export const { t } = i18n

export default i18n
