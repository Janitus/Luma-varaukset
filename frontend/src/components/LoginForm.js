import { useMutation } from '@apollo/client'
import React, { useEffect } from 'react'
import { useHistory } from 'react-router'
import { LOGIN } from '../graphql/queries'
import { useField } from '../hooks'

const LoginForm = ({ getUser, sendMessage }) => {
  const history = useHistory()
  const username = useField('')
  const password = useField('password')
  const [login, result] = useMutation(LOGIN, {
    onError: (error) => sendMessage(error.graphQLErrors[0].message, 'danger'),
  })

  useEffect(() => {
    if (result.data) {
      const token = result.data.login.value
      localStorage.setItem('app-token', token)
      getUser()
      history.push('/')
    }
  }, [result.data])

  const submit = (event) => {
    event.preventDefault()
    login({
      variables: {
        username: username.field.value,
        password: password.field.value,
      },
    })
    username.clear()
    password.clear()
  }

  const cancel = (event) => {
    event.preventDefault()
    history.push('/')
  }

  const style = { width: 500 }
  return (
    <div className="container">
      <div className="columns is-centered">
        <div className="section">
          <div className="title">Kirjautuminen (Luma-Varaukset)</div>
          <form onSubmit={submit}>
            <div className="field">
              <label className="label">Käyttäjänimi</label>
              <div className="control">
                <input
                  id="username"
                  type="text"
                  className="input"
                  {...username.field}
                  style={style}
                />
              </div>
            </div>
            <div className="field">
              <label className="label">Salasana</label>
              <div className="control">
                <input
                  id="password"
                  type="password"
                  className="input"
                  {...password.field}
                  style={style}
                />
              </div>
            </div>
            <div className="field is-grouped">
              <div className="control">
                <button id="login" className="button luma primary" type="submit">
                  Kirjaudu sisään
                </button>
              </div>
              <div className="control">
                <button className="button luma" onClick={cancel}>
                  Poistu
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  )
}

export default LoginForm
