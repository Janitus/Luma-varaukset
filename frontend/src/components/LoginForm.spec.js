import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { render, waitFor, fireEvent, screen } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import { Router } from 'react-router-dom'
import { MockedProvider } from '@apollo/client/testing'
import { GraphQLError } from 'graphql'
import { LOGIN } from '../graphql/queries'
import '../i18n'

import LoginForm from './LoginForm'

const loginSuccess = [
  {
    request: {
      query: LOGIN,
      variables: { username: 'Test', password: 'Test' }
    },
    result: {
      data: {
        login: { 'value': 'abrakadabra' },
      },
    },
  },
]

const loginFailed = [
  {
    request: {
      query: LOGIN,
      variables: { username: 'Test', password: 'Test' }
    },
    result: {
      errors: [new GraphQLError('Error!')]
    },
  },
]

test('Renders without error', () => {
  const { container } = render(
    <MockedProvider mocks={loginSuccess} addTypename={false}>
      <LoginForm/>
    </MockedProvider>
  )
  expect(container.firstChild).toMatchSnapshot()
})

test('Login successful', async () => {
  let mockStorage = {}
  const setItemSpy = jest
    .spyOn(global.Storage.prototype, 'setItem')
    .mockImplementation((key, value) => {
      mockStorage[key] = value
    })
  const history = createMemoryHistory()
  const getUser = jest.fn()
  const { container } = render(
    <MockedProvider mocks={loginSuccess} addTypename={false}>
      <Router history={history}>
        <LoginForm getUser={getUser}/>
      </Router>
    </MockedProvider>
  )
  const usernameInput = container.querySelector('#username')

  fireEvent.change(usernameInput, {
    target: { value: 'Test' }
  })
  const passwordInput = container.querySelector('#password')
  fireEvent.change(passwordInput, {
    target: { value: 'Test' }
  })
  const loginButton = await container.querySelector('#login')
  fireEvent.click(loginButton)
  await waitFor(() => new Promise((res) => setTimeout(res, 300))) // Allow component time to render
  expect(getUser).toHaveBeenCalled()
  expect(setItemSpy).toHaveBeenCalled()
  expect(history.location.pathname).toBe('/')
  expect(mockStorage['app-token']).toEqual('abrakadabra')
  setItemSpy.mockRestore()
})

test('Login not successful', async () => {
  const history = createMemoryHistory()
  const getUser = jest.fn()
  const sendMessage = jest.fn()
  const { container } = render(
    <MockedProvider mocks={loginFailed}>
      <Router history={history}>
        <LoginForm getUser={getUser} sendMessage={sendMessage} />
      </Router>
    </MockedProvider>
  )
  const usernameInput = container.querySelector('#username')

  fireEvent.change(usernameInput, {
    target: { value: 'Test' }
  })
  const passwordInput = container.querySelector('#password')
  fireEvent.change(passwordInput, {
    target: { value: 'Test' }
  })
  const loginButton = await container.querySelector('#login')
  fireEvent.click(loginButton)
  await waitFor(() => new Promise((res) => setTimeout(res, 300))) // Allow component time to render
  expect(sendMessage).toHaveBeenCalledTimes(1)
})

test('Return on clicking "Back"', async () => {
  const history = createMemoryHistory()

  const { container } = render(
    <MockedProvider mocks={loginSuccess} addTypename={false}>
      <Router history={history}>
        <LoginForm/>
      </Router>
    </MockedProvider>
  )
  const backButton = await container.querySelector('button:not(.primary)')
  fireEvent.click(backButton)
  //await waitFor(() => new Promise((res) => setTimeout(res, 0))) // Allow component time to render
  expect(history.location.pathname).toBe('/')
})

