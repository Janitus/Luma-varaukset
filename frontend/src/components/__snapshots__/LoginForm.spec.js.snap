// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without error 1`] = `
<div
  class="container"
>
  <div
    class="columns is-centered"
  >
    <div
      class="section"
    >
      <div
        class="title"
      >
        Kirjautuminen (Luma-Varaukset)
      </div>
      <form>
        <div
          class="field"
        >
          <label
            class="label"
            for="username"
          >
            Käyttäjätunnus
          </label>
          <div
            class="control"
          >
            <input
              autocomplete="username"
              class="input undefined"
              id="username"
              name="username"
              type="input"
              value=""
            />
          </div>
        </div>
        <div
          class="field"
          style="width: 500px;"
        >
          <label
            class="label"
            for="password"
          >
            Salasana
          </label>
          <div
            class="control"
          >
            <input
              autocomplete="current-password"
              class="input undefined"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          class="field is-grouped"
        >
          <div
            class="control"
          >
            <button
              class="button luma primary"
              id="login"
              type="submit"
            >
              Kirjaudu sisään
            </button>
          </div>
          <div
            class="control"
          >
            <button
              class="button luma"
            >
              Poistu
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
